ID: 50316
Title: Unexpected error (Unknown identifier: 0) when using bitOr and width_bucket
Description:
**Describe what's wrong**

The two semantically-equivalent SELECT statements should output the same results, but one of them throws an exception.

**Does it reproduce on recent release?**

It can be reproduced in the latest version.

**How to reproduce**

Version: 23.5.1 (806ef08e5f6b3fc93993d19aac505290e9bfa554)

Easy reproduce in ClickHouse fiddle: https://fiddle.clickhouse.com/fbee9460-ce40-4e28-8998-38a2c84ca645

_Set up database_

```sql
create table t1 (c12 UInt32)engine = TinyLog;
insert into t1 values (1);
```

_SELECT statement 1_

```sql
with cte_1 as (
  select  
    subq_2.c_5_c720_6 as c_2_c727_1
  from 
    (select  
          count(1) as c_5_c717_3, 
          sum(1) as c_5_c720_6
        from 
          t1 as subq_1
        group by subq_1.c12 having true) as subq_2
  where (case when (subq_2.c_5_c717_3 <> subq_2.c_5_c717_3) then '1' else '1' end) = (case when (0 in (select 1)) then '1' else '1' end))
select *
  from cte_1 as ref_13
  where (bitOr((96) * (-0), 1) + width_bucket(0, 0, ref_13.c_2_c727_1, 1)) between 0 and 10000;
```

As `(case when (subq_2.c_5_c717_3 <> subq_2.c_5_c717_3) then '1' else '1' end)` can be replaced with `'1'`, I get the semantically-equivalent SELECT statement:

_SELECT statement 2_

```sql
with cte_1 as (
  select  
    subq_2.c_5_c720_6 as c_2_c727_1
  from 
    (select  
          count(1) as c_5_c717_3, 
          sum(1) as c_5_c720_6
        from 
          t1 as subq_1
        group by subq_1.c12 having true) as subq_2
  where '1' = (case when (0 in (select 1)) then '1' else '1' end))
select *
  from cte_1 as ref_13
  where (bitOr((96) * (-0), 1) + width_bucket(0, 0, ref_13.c_2_c727_1, 1)) between 0 and 10000;
```

**Expected behavior**

The two SELECT statements output the same results.

**Actual behavior**

They are different.

SELECT statement 1 throws an exception:
```
Received exception from server (version 23.5.1):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Unknown identifier: 0: While processing ((if((count() AS c_5_c717_3) != c_5_c717_3, '1', '1') = if(0 IN ((SELECT 1) AS _subquery13), '1', '1')) AND ((bitOr(96 * 0, 1) + widthBucket(0, 0, count() AS c_5_c720_6, 1)) >= 0)) AND ((bitOr(96 * 0, 1) + widthBucket(0, 0, c_5_c720_6, 1)) <= 10000). (UNKNOWN_IDENTIFIER)
(query: with cte_1 as (select  
    subq_2.c_5_c720_6 as c_2_c727_1
  from 
    (select  
          count(1) as c_5_c717_3, 
          sum(1) as c_5_c720_6
        from 
          t1 as subq_1
        group by subq_1.c12 having true) as subq_2
  where (case when (subq_2.c_5_c717_3 <> subq_2.c_5_c717_3) then '1' else '1' end) = (case when (0 in (select 1)) then '1' else '1' end))
select *
  from 
    cte_1 as ref_13
  where (bitOr((96) * (-0), 1) + width_bucket(0, 0, ref_13.c_2_c727_1, 1)) between 0 and 10000;)
```

SELECT statement 2 outputs:
```
+------------+
| c_2_c727_1 |
+------------+
|          1 |
+------------+
```

**Additional context**

The earliest reproducible version is 23 in fiddle: https://fiddle.clickhouse.com/c672e8a6-8372-4e2e-a688-ec566654aa6b

